File: content/08-heap-priority-queue/top_k_frequent_elements.py
from collections import Counter

def topKFrequent(nums, k: int):
    count = Counter(nums)
    freq = [[] for _ in range(len(nums) + 1)]

    for n, c in count.items():
        freq[c].append(n)

    res = []
    for i in range(len(freq) - 1, 0, -1):
        for n in freq[i]:
            res.append(n)
            if len(res) == k:
                return res
    return res

if __name__ == "__main__":
    assert sorted(topKFrequent([1,1,1,2,2,3], 2)) == [1,2], "Test 1 Failed"
    assert topKFrequent([1], 1) == [1], "Test 2 Failed"
    print("All tests passed!")

File: content/08-heap-priority-queue/top_k_frequent_elements.js
function topKFrequent(nums, k) {
    let count = {};
    for (let n of nums) count[n] = (count[n] || 0) + 1;

    let freq = Array.from({ length: nums.length + 1 }, () => []);
    for (let n in count) freq[count[n]].push(Number(n));

    let res = [];
    for (let i = freq.length - 1; i > 0; i--) {
        for (let n of freq[i]) {
            res.push(n);
            if (res.length === k) return res;
        }
    }
    return res;
}

const assert = require('assert');
assert.deepStrictEqual(topKFrequent([1,1,1,2,2,3], 2).sort(), [1,2]);
assert.deepStrictEqual(topKFrequent([1], 1), [1]);
console.log("All tests passed!");
